### ISSUE TITLE
Support for validating modifications to the data structures

### ISSUE THREAD

**Issue by hackaugusto** [0xPolygonMiden/crypto#222]
When applying changes sent from the network, the following steps should be followed

- temporarily apply the changes
- validate the result
- commit the changes

The above is necessary for two cases:

- Identify bugs, for cases were there are data corruption or bugs in the serde logic
- Prevent attacks, for cases were the input is not trusted

The issue is that most of the datastructures currently have a single version, with the exception of the `MerkleStore`.

Possible solutions:

- Move the logic of the existing datastructure to the merkle store. The merkle store can be modified in place since it preserves the old versions in memory too
- As suggested by @frisitano , we can add a wrapper around the existing datastructures, which allows these modifications to be done without changing the underlying data, and commiting it once the results have been validated.
    - I believe it would be useful to have some sort of trait abstraction for this https://github.com/0xPolygonMiden/crypto/issues/221 , so that one generic wrapper implementation can be used with the many structures we currently have.

**Comment by hackaugusto** [comment]
The motivation behind this issue is the `apply_block` endpoint. If it receives a block header with the root of the nullifier tree and the nullifiers, it should verify that the two match.

**Comment by frisitano** [comment]
> * As suggested by @frisitano , we can add a wrapper around the existing datastructures, which allows these modifications to be done without changing the underlying data, and commiting it once the results have been validated.
>   
>   * I believe it would be useful to have some sort of trait abstraction for this [Trait abstraction to interact with existing data structures #221](https://github.com/0xPolygonMiden/crypto/issues/221) , so that one generic wrapper implementation can be used with the many structures we currently have.

My thinking is that the data structures should have a database transaction like mechanic in which changes are pushed into separate storage overlay and only when `.commit()` is called should they be persisted into the data structures main store. If we need to rollback the changes then we can call `.rollback()`. This is quite a large undertaking and as such I think the proposal of cloning the data structures is fine for now since the volumes of data we are dealing with is relatively small.

**Comment by bobbinth** [comment]
I wonder if we could combine this idea with the idea @hackaugusto had earlier to remove hash computations from the "critical path" of `apply_block()` endpoint (specifically, so that we need to lock the database for writes only for a very short period of time).

On an example of account tree, it could work like so:
- We apply updates to the tree without committing them. This step will perform all required hashing and in the end, we can verify that the resulting Merkle root matches the expected one.
- We lock the database for write and commit the updates. This modifies the account tree but does not require performing any hashing. Thus, the lock time should be pretty short.

One potential way to do this (on an example of the `SimpleSmt`) could be as follows:

We can add a method to the `SimpleSmt` which would look something like this:
```Rust
pub fn build_changeset(&self, updates: Vec<(u64, Word)>) -> Result<ChangeSet, MerkleError>
// I don't love the name "changeset" - so, open to suggestions here
```
This method would perform all the relevant hashing and store or results in the `ChangeSet` struct which would also contain the new root of the tree. Using this changeset, we'd be able to check the validity of the root.

Then, we'd be able to apply the changeset to the tree like so:
```Rust
pub fn apply_changes(&mut self, changes: ChangeSet) -> Result<(), MerkleError>
```
This method would update all the internal data structures of the SMT based on the provided change set (ideally, by just moving data in memory).

**Comment by bobbinth** [comment]
Closed by #327.
